tracee-rules/signatures/golang/stdio_over_socket.go

```diff
@@ -0,0 +1,159 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/aquasecurity/tracee/tracee-rules/types"
+	tracee "github.com/aquasecurity/tracee/tracee/external"
+	"strings"
+)
+
+type connectAddrData struct {
+	SaFamily string `json:"sa_family"`
+	SinPort  string `json:"sin_port"`
+	SinAddr  string `json:"sin_addr"`
+}
+
+type stdIoOverSocket struct {
+	cb              types.SignatureHandler
+	processSocketIp map[int]map[int]string
+}
+
+func (sig *stdIoOverSocket) Init(cb types.SignatureHandler) error {
+	sig.cb = cb
+	sig.processSocketIp = make(map[int]map[int]string)
+
+	return nil
+}
+
+func (sig *stdIoOverSocket) GetMetadata() (types.SignatureMetadata, error) {
+	return types.SignatureMetadata{
+		Name:        "STD I/O Over Socket",
+		Description: "Redirection of process's standard input/output to socket",
+		Tags:        []string{"linux", "container"},
+		Properties: map[string]interface{}{
+			"Severity":     3,
+			"MITRE ATT&CK": "Persistence: Server Software Component",
+		},
+	}, nil
+}
+
+func (sig *stdIoOverSocket) GetSelectedEvents() ([]types.SignatureEventSelector, error) {
+	return []types.SignatureEventSelector{
+		{Source: "tracee", Name: "connect"},
+		{Source: "tracee", Name: "dup2"},
+		{Source: "tracee", Name: "close"},
+	}, nil
+}
+
+func (sig *stdIoOverSocket) OnEvent(e types.Event) error {
+
+	eventObj, ok := e.(tracee.Event)
+	if !ok {
+		return fmt.Errorf("invalid event")
+	}
+
+	var connectData connectAddrData
+	socketIpMap := make(map[int]string)
+	stdAll := []int{0, 1, 2}
+
+	pid := eventObj.ProcessID
+
+	switch eventObj.EventName {
+
+	case "connect":
+
+		sockfdArg, err := GetTraceeArgumentByName(eventObj, "sockfd")
+		if err != nil {
+			return err
+		}
+
+		sockfd := sockfdArg.Value.(int)
+
+		addrArg, err := GetTraceeArgumentByName(eventObj, "addr")
+		if err != nil {
+			return err
+		}
+
+		addrStr := strings.Replace(addrArg.Value.(string), "'", "\"", -1)
+		err = json.Unmarshal([]byte(addrStr), &connectData)
+		if err != nil {
+			return fmt.Errorf(err.Error())
+		}
+
+		if connectData.SaFamily == "AF_INET" {
+			socketIpMap[sockfd] = connectData.SinAddr
+			sig.processSocketIp[pid] = socketIpMap
```

Comment:
this is the expected behavior because we're looking for a dup call that will duplicate oldfd into newfd. oldfd is the currently existing FD for this pid, and therefore this assignment keeps us up to date with the current open socket FDs.

Reply:
Not sure I understand what you mean.
What about the following case:
```
connect(5,...)
connect(2,...)
dup2(5,0)
```
In this case, fd 5 is connected to socket and stdin, but the signature will not hold, am I wrong?